### ISSUE TITLE
How to reconnect after: ConnectionClosed: code 1006 <connection closed abnormally> ?

### ISSUE THREAD

**Issue by ShervinDD** [aaugustin/websockets#367]
Hi, I'm still new to python and websockets and can't figure this out.

I am connected to two websockets. After a while, one of the websockets disconnects and gives ConnectionClosed: Websockets connection is closed: code = 1006 <connection closed abnormally [internal]>, no reason

I don't know how to fix this other than to detect it with "if not ws.open:" and then reconnect. But I can't seem to get it to reconnect to the websocket.

Here is the websockets code I'm using, and below this is how I tried to get it to reconnect but failed:

```
import asyncio
import json
import websockets

async def socket_connect_ws():
    async with websockets.connect('wss://ws.pusherapp.com/app/de504dc5763aeef9ff52?protocol=7') as ws:
        msg = json.dumps({"event":"pusher:subscribe","data":{"channel":"order_book"}})
        await ws.send(msg)
        while True:
            try:
                packet = await ws.recv()
                if 'bids' in packet:
                    print(json.loads(json.loads(packet)['data'])['bids'][0])
                    await asyncio.sleep(1)
            except Exception as e:
                print('Error!: ', e)
                break
loop = asyncio.get_event_loop()
asyncio.ensure_future(socket_connect_ws())
loop.run_forever()
```

Now trying to get it to detect websocket closed (after error 1006) and reconnect, but this doesn't work.

```
import asyncio
import json
import websockets

async def socket_connect_ws():
    async with websockets.connect('wss://ws.pusherapp.com/app/de504dc5763aeef9ff52?protocol=7') as ws:
        msg = json.dumps({"event":"pusher:subscribe","data":{"channel":"order_book"}})
        await ws.send(msg)
        while True:
            if not ws.open:
                print ('Websocket NOT connected. Trying to reconnect.')
                async with websockets.connect('wss://ws.pusherapp.com/app/de504dc5763aeef9ff52?protocol=7') as ws:
                    msg = json.dumps({"event":"pusher:subscribe","data":{"channel":"order_book"}})
                    await ws.send(msg)
            try:
                packet = await ws.recv()
                if 'bids' in packet:
                    print(json.loads(json.loads(packet)['data'])['bids'][0])
                    await asyncio.sleep(1)
            except Exception as e:
                print('Error!: ', e)
                break
loop = asyncio.get_event_loop()
asyncio.ensure_future(socket_connect_ws())
loop.run_forever()
```

**Comment by aaugustin** [comment]
> Now trying to get it to detect websocket closed (after error 1006) and reconnect, but this doesn't work.

Could you clarify how it fails?

One possible issue:

```python
                async with websockets.connect('wss://ws.pusherapp.com/app/de504dc5763aeef9ff52?protocol=7') as ws:
                    msg = json.dumps({"event":"pusher:subscribe","data":{"channel":"order_book"}})
                    await ws.send(msg)
```

will close the connection as soon as you exit the `async with` block, which I suspect isn't what you intend.

**Comment by ShervinDD** [comment]
I made a mistake. After the ConnectionClosed error code 1006, it detects the websocket is closed via the ```if not ws.open:``` statement, and the websocket reconnects. However, it does not return to the original while loop and ```print(json.loads(json.loads(packet)['data'])['bids'][0])```

You are correct about wanting the connection to stay open at all times. How do I get the websocket to stay open after it detects a disconnect?

**Comment by ShervinDD** [comment]
Looks like I got it to work by removing the ```async with``` as you suggested:

```Python
if not ws.open:
                print ('Websocket NOT connected. Trying to reconnect.')
                ws = await websockets.connect('wss://ws.pusherapp.com/app/de504dc5763aeef9ff52?protocol=7')
                msg = json.dumps({"event":"pusher:subscribe","data":{"channel":"order_book"}})
                await ws.send(msg)
```
Thank you @aaugustin for your help on this!

**Comment by aaugustin** [comment]
Good to hear your issue is resolved.

You could also start an asyncio task that calls `ws.ping()` regularly. That may help keeping the connection open.

**Comment by mgavaudan** [comment]
Hi I have the same exact issue but your solution doesn't seem to be working.

`    
async def get_order_book(symbol):
    with open('test.csv', 'a+') as csvfile:
        csvw = csv.writer(csvfile, delimiter=',', quotechar='|', quoting=csv.QUOTE_MINIMAL)
        DT = Data(data=data, last_OB_id=ob_id, last_TR_id=tr_id, sz=10, csvw=csvw)
        while True:
            if not websocket.open:
                print('Reconnecting')
                websocket = await websockets.connect(ws_url)
            else:
                resp = await websocket.recv()
                update = ujson.loads(resp)
                DT.update_data(update)

async def get_order_books():
    r = requests.get(url='https://api.binance.com/api/v1/ticker/24hr')
    await asyncio.gather(*[get_order_book(data['symbol']) for data in r.json()])

if __name__ == '__main__':
    asyncio.run(get_order_books())
`

The way I've been testing it is by closing my internet connection, but after a ten second delay it still returns the 1006 error.

I'm running Python 3.7 and Websockets 7.0.

Let me know what your thoughts are, thanks!

**Comment by aaugustin** [comment]
My thoughts are https://websockets.readthedocs.io/en/stable/contributing.html#bitcoin-users

**Comment by mgavaudan** [comment]
@aaugustin Proof of stake is coming, soon proof of work (which consumes the energy you're worried about) will be obsolete. Regardless, having a free financial system with no quantitative easing is much more important than consuming a little electricity. Furthermore, it's been shown that gold mining still uses much more electricity than Bitcoin, would you also boycott any question related to that or any other industry that consumes vast amounts of energy?

**Comment by aaugustin** [comment]
Yes, I'm aware of all this, and no, I'm not getting into this discussion — I hope you understand ;-)

**Comment by asohn123** [comment]
Hi @aaugustin, im doing a Kaggle project with geological data and ran into the same problem as @mgavaudan (the one cited above). Are you aware of any solutions to it?

**Comment by aaugustin** [comment]
Could you file a new issue with an explanation of your problem and, ideally, an example that allows reproducing it?

**Comment by esundberg** [comment]
I tried this and it seems to be working

```
# Connection Loop to handle 1st connecting and reconnects
    while True:
        try:
            if ws.open:
                print(f'Web Socket is open')
        except:
            print(f'Websocket is not open... Trying to connect or reconnect')
            #print(f'Connecting to WS')
            ws = await websockets.connect(uri)
        else: 
            print(f'Websocket open waiting for rx msg')
            msg = await ws.recv()
            result = await on_message(msg)
```

**Comment by yedhrab** [comment]
Make `ping_interval=None` in ws connection so connect like `websockets.connect(config.WS_BTCTURK_URL, ssl=True, ping_interval=None)`
> Check out this  stackoverflow link [👪 1006 Connection closed abnormally error with python 3.7 websockets](https://stackoverflow.com/a/58993145/9770490)

**Comment by ShervinDD** [comment]
A potential solution for this is removing `await asyncio.sleep(1)` entirely, or reducing the sleep time to something small like `await asyncio.sleep(0.0001)`. My hypothesis is high sleep times result in a large build-up of messages, which causes the 1006 error. 

Please let me know if this solves the issues for anyone.

**Comment by lgrahl** [comment]
That is most certainly not the cause. TCP flow control would kick in in such a case.

**Comment by mrajancsr** [comment]
Hello, i am getting an issue with this code.  
Currently im constantly seeing this message: 
INFO:root:exception occured at ftx-blotter
INFO:root:Socket Closed: at 2022-03-10 18:07:44.202593;
            close reason: ,
                close code: 1006
 
It seems im getting disconnected after couple hours when i step away from my computer.  
``` await obj.send()

    while True:
        try:
            records = await obj.receive()
        except Exception as e:
            obj.logger.info(f"exception occured at {stream_name}")
            msg = f"""Socket Closed: at {datetime.now()};
            close reason: {obj.websocket.close_reason},
                close code: {obj.websocket.close_code}"""
            print(msg)
            obj.logger.info(msg)
            obj.logger.info("\n")
            obj.logger.info(e)
            print("reconnecting...")
            obj.websocket = await websockets.connect(obj.url, ping_interval=None)
            await obj.send()
            continue
        if not records:
            continue

        record_count += len(records)```
  Above is a partial snapshot of my code, 
  I enter the code as such, 
  async with Blotter('btc/usd', 'binance-blotter') as blotter:
          await blotter.stream()
  
The Blotter class encapsulates the websockets class.  
Inside stream(), i have devined the snapshot code as above.  
Why is this error constantly occuring?  I have deployed 10 currencies for 4 exchange to get 2 different types of data, orderbook and blotter.  This amounts to 80 processes/currencies that i am streaming data for.  
1) Is this a computer issue that would be resolved if deployed on aws? 
2) If not, is this a code issue with the way instantiating the class with the ```async with``` keyword?  Another thing is, with these 80 different but independent processes, if one fails, they all fail.  Why does this happen? i get two types of errors, sometimes, where close_reason is None and sometimes when close_reason is 1006.  Thanks
Thanks

**Comment by Futuristic263** [comment]
https://pythondev.space/questions/proceed-after-an-attempt-with-no-success



I realized the answer
